### ISSUE TITLE
GraphQL LSP, Codemirror & Monaco Roadmap [WIP]

### ISSUE THREAD

**Issue by acao** [graphql/graphiql#983]
# GraphQL IDE Ecosystem Roadmap

![Screen Shot 2019-11-25 at 1 19 22 PM](https://user-images.githubusercontent.com/49038/69566927-3821e080-0f86-11ea-94d1-629ba9648d6c.png)

This is a WIP. Hoping to add more items. But it shows the trajectory of where everything needs to go in our biggest undertaking, to adopt an LSP plugin system, a react enabled plugin system for GraphiQL that layers over that, and of course monaco. New features will need to fit at the end of this roadmap, essentially.

## GraphiQL

We want to take the current GraphiQL - make it solid, modern and be prepared for the re-design.

`0.x.y`

- [X] - #977 - full i18next!
- [X] - #967 - full accessibility!
- [ ] - e2e harness with Cypress and a nice suite to start with
- [ ] - React 16! Use deprecated class methods for now
- [ ] - Get smarter with a core react context, provide `useGraphQLContext` hook, maybe with parsing and visitors/ast?
- [ ] - Expose events interface to use internally and via plugins via `useReducer`, etc?
- [ ] - Convert several smaller components to implement the hooks interface, use for existing functional components
- [ ] - Begin with a 'plugins' array prop that renders Components that use context/reducer events and/or FCs that use hooks interface

Then we want to switch the editor to monaco:

`1.0.0`
- [ ] - Adopt new monaco language mode, stabilize
- [ ] - keymap/command customization to allow previous (sublime based) keymap, or new vscode-oriented keymap, or per-shortcut customization. Default keymap should have parity with last 0.x.y release?

## GraphQL Language Service Protocol (LSP)
- `graphql-language-interface`
- `graphql-language-parser`

With monaco, we can start working with a Language Server Protocol, which can be consumed by everyone on the web and in their IDE.

- [X] - convert to TS
- [ ] - Symbols
- [ ] - Go To Reference
- [ ] - Plugins in the interface?
- [ ] - Adopt upcoming 15.0.x parser capabilities
- [ ] - Replace `@babel/parser` with a language agnostic parser for `gql` template tags and other configurable patterns.
- [ ] - Universal file system
- [ ] - Universal config, graphql-config?

## GraphQL Service 
- `graphql-language-service`
- `graphql-language-service-server`

The actual server and service itself, that consumes the LSP for vscode, other IDE language servers. Basically the monaco <-> graphql-lsp sync.

- [ ] - convert to TS
- [ ] - upgrade `vscode-languageserver` version several major versions !!

## Monaco GraphQL
- ` monaco-graphql` (new)
- ` examples/monaco-example` (new)
- `examples/monaco-webpack-example` (new)
New! We are working towards a stable monaco mode that will become a part of the official, stable GraphiQL reference implementation. There is a fairly old branch where we have already begun on this work and the example implementation, called `monaco-test`.

Currently what's left:
- [ ] - variables JSON completion/diagnostics/declarations etc (dynamically define json schema, so much simpler than with codemirror!)
- [ ] - launch language service(s) in webworker(s) (see: monaco-json, monaco-typescript, etc)
- [ ] - re-register default `graphql` language with our own version somehow? probably patterns for this...
- [ ] - Declarations support  
  - [ ] - instantiate the entire `GraphQLLanguageService` (interface pkg) class to instantiate the service
  - [ ] - create `GraphQLCache` instance
- [ ] - find a way to contextually load variables json or results json or even AST json?

## Codemirror GraphQL

Should be officially maintained at least as long as `0.x.y` versions of `GraphiQL` are supported.
Ideally should live on much longer.

- [ ] - Codemirror 6 is on the way?
- [ ] - Mounting the `GraphQLLanguageServiceInterface` class here as well? Looks like theres an LSP compatible interface for GraphQL?
- [x] - Accessibility improvements
- [ ] - Convert to TS?
- [ ] - Convert tests to Jest?

**Comment by orta** [comment]
Want to help out? Reach out and say hi in [the discord](https://discord.gg/eNuu9Cb)

**Comment by acao** [comment]
@orta thanks for all your awesome updates and the graphic! I'm going to add some more items and start cross-referencing issues in the queue today.  See, about starting to break things up!

**Comment by acao** [comment]
there is still so much to add to this! @orta had the great idea of breaking out, which we’ve done with github projects. in fact, we just closed a project for the first time! (Webpack Takeover)

**Comment by moulikcipherX** [comment]
Hello @orta I would love to work on graphql service by starting to convert in Typescript.

**Comment by acao** [comment]
@moulikcipherX good news, we've already converted it to typescript! about to schedule new working group schedule and Plugin API spec proposals here soon!

@orta re-opening this to turn it into a mega-issue, adding plugins as well?

**Comment by acao** [comment]
@orta @benjie what do y'all think now?

**Comment by acao** [comment]
> Hello @orta I would love to work on graphql service by starting to convert in Typescript.

graphql language service conversion was completed last year, but thank you though!

**Comment by vojtapol** [comment]
Any updates on this?

**Comment by acao** [comment]
plugins made 2.0 architecture too complicated for the next major version, and also I'm no longer being paid to work on this monorepo as of this fall, so the roadmap has changed based on my availability.

because I'm one of the only active maintainers for this ecosystem, I mostly only have time for the minimum - maintenance work, advancing new GraphQL features so we can advance the spec, and ensuring the LSP ecosystem supports all GraphQL features fully.

[there is a WIP](https://github.com/graphql/graphiql/tree/main/packages/graphiql-2-rfc-context) for 2.0 since early this year. it was merged to master branch, then we moved it to a seperate workspace because of performance issues with context. the readme will be updated with more details soon, and I've been trying to recruit folks to offer new proposals to finish 2.x and no one has opened a PR. i have a lot more community/devrel work to do. was hoping for hacktober, but at work I was pivoted to work on wordpress plugins well before I could make that deadline.

most of the time I spent being paid to work on this repo, I spent on `monaco-graphql` which is a requirement for 2.0, supporting graphiql@1.0, and advancing more complete LSP support for 2.0.

Instead of plugins, for 2.0, we will just have fantastic components, hooks and state implementation (an SDK so to speak, for advanced custom implementations) and an optional top-level service layer GraphiQL component and middlewares. Hopefully sometime soon I can re-hash all the architectural plans, but at this point, there are too many pressing issues and features to address in this repo outside of the `graphiql` workspace.

The goal here is not plugins, but enhanced customizeability and extensibility. Since we already need to rewrite the entire clientside domain model with a new state management tool (yet to be chosen), adding the complexity of an entire plugin API isn't work it, when people can just publish react components and libraries that can be imported to new and powerful component props.

If you want a full-featured, open source client that uses GraphQL and supports a rich plugin ecosystem *now*, Altair and Insomnia are both excellent choices for this.

If you want to build your own tool with a plugin ecosystem with react, svelte, vue, plain js, whatever you choose, check out the github api explorer demo we have for `monaco-graphql`!

Then, this 2.0 SDK will change for 3.0 to some degree, and potentially a plugin service layer will be introduced

**Comment by acao** [comment]
I've got a really exciting "fork" on this proposal for y'all here:

https://github.com/graphql/graphiql/discussions/1787

we're also writing a WIP roadmap
https://docs.google.com/document/d/1hFJ2DgmzmeE7Lr5cdgT3srm89_VgR3dEGyr9gLqPQJk/edit?usp=sharing

and yes, anyone is able to edit it!

**help with roadmapping**

if you want to help advance the roadmapping effort, and help reduce the issue count at the same time, look for issues labelled "plugin", or at any other feature requests, and add them to the GraphiQL 2.0 roadmap. you may need to formulate a summary of the issue and what it represents.

once an feature request has been added to the roadmap, you can add @acao #roadmapped at the end of the comment, and I can do a strict issue search for all of them. this would help me so much! 

we're open to any improvements to this process
